augur/dates/__init__.py

```diff
@@ -16,6 +16,18 @@
     3. a backwards-looking relative date in ISO 8601 duration format with optional P prefix (e.g. '1W', 'P1W')
 """)
 
+
+# Matches floats (e.g. 2018.0, -2018.0).
+# Note that a year-only value is treated as incomplete ambiguous and must be
+# non-negative (see RE_YEAR_ONLY).
+RE_NUMERIC_DATE = re.compile(r'^-?[0-9]*\.[0-9]*$')
```

Comment:
Matching float literals is complex; this will only match a subset of them. I think worth considering what we're excluding in a move from `float(x)` to this regex. For example, `float("+2022.1")` and `float("2.0221e3")` and `float("2_022.1")` all work, but won't be recognized by this regex, and so will never be tried when gating on the regex matching.

Broadly, I think the choice is between 1) keep accepting ~anything `float()` parses for better interop (e.g. lots of things produce the `XeY` notation) and don't attempt to regex match, or 2) explicitly narrow what we accept to a much smaller syntax subset than `float()` and use the regex to gate it.

Separately, this regex will match the strings `.` and `-.`, which seems wrong (although the subsequent `float()` will fail anyway).

Reply:
Thanks for raising this point. The motivation was to avoid letting year-only dates go through `float()`. I think option (1) of above sounds like the better choice in the short term. The motivation can be satisfied by changing the order of operations:

```py
if RE_YEAR_ONLY.match(date):
    date = incomplete_to_ambiguous(fmt, year=date)

try:
    return float(date)
except ValueError:
    pass
```

Reply:
Sounds good to me!